wporg-two-factor.php

```diff
@@ -33,3 +38,114 @@ function two_factor_providers( array $providers ) : array {
 
 	return array_intersect_key( $providers, $desired_providers );
 }
+
+
+/**
+ * Remove a user's Super Admins status if they don't have 2FA enabled.
+ *
+ * This is needed in addition to `remove_capabilities_until_2fa_enabled()` for two reasons:
+ *     1: To protect against code that calls `is_super_admin()` directly, instead of checking capabilities.
+ *     2: To avoid the code in `has_cap()` that allows Super Admins to do anything unless `do_not_allow` is set.
+ *        That would interfere with reducing their capabilities to a Subscriber in `remove_capabilities_until_2fa_enabled()`.
+ */
+function remove_super_admins_until_2fa_enabled() : void {
+	global $super_admins;
+
+	$user     = wp_get_current_user();
+	$position = array_search( $user->user_login, $super_admins, true );
+
+	if ( false === $position ) {
+		return;
+	}
+
+	if ( user_requires_2fa( $user ) && ! Two_Factor_Core::is_user_using_two_factor( $user->ID ) ) {
+		unset( $super_admins[ $position ] );
+	}
+}
+
+/**
+ * Remove capabilities when a user with elevated privileges hasn't enabled 2FA.
+ *
+ * That is necessary even though we'll redirect all requests to their profile, because otherwise they could still
+ * perform privileged actions on the front end, via the REST API, etc.
+ */
+function remove_capabilities_until_2fa_enabled( array $allcaps, array $caps, array $args, WP_User $user ) : array {
+	if ( 0 === $user->ID || ! user_requires_2fa( $user ) ) {
+		return $allcaps;
+	}
+
+	if ( ! Two_Factor_Core::is_user_using_two_factor( $user->ID ) ) {
+		// This also relies on `remove_super_admins_until_2fa_enabled()`, see notes in that function.
+		$allcaps = array(
+			'subscriber' => true,
+			'read'       => true,
+		);
```

Comment:
Another option here I think is to return the `do_not_allow` cap when the requested cap is anything other than the whitelisted caps.

I believe that would override the super-admin always-truthful caps too, but `remove_super_admins_until_2fa_enabled()` would be best kept still then too.

Reply:
Ah, that's a good point 👍🏻

Reply:
Er, nevermind, it looks like `user_has_cap` doesn't even get called in that case.

https://github.com/WordPress/wordpress-develop/blob/a71d7373ad6ce0a8f252717832f42aebc3e08854/src/wp-includes/class-wp-user.php#L780-L809